### ISSUE TITLE
Current note editor is not accessible to screenreaders

### ISSUE THREAD

**Issue by zersiax** [archivy/archivy#205]
Unfortunately, the editor to compose note contents is not accessible using screen readers at present. From the behavior I'm seeing I think it's either Ace or CodeMirror, which unfortunately don't work well with assistive technology.
Please consider alternatives like Monaco, or an alternative where screen reader users are able to use a standard <textarea> or the contenteditable attribute in order to fix this, as at present it is a showstopper and prevents people using screenreaders such as myself to use Archivy the way it was meant to be used.
Temporary workaround is to write note contents somewhere else and copy-paste it into the field, which works, but is far from ideal.
Awesome concept for the rest, and thanks a lot for reading at least! :)

**Comment by ddoyaguez** [comment]
Hi. I'm checking this with Orca screen reader in Linux with Firefox and I'm able to write text into the easyMDE editor, that is based on CodeMirror, but screen reader is not able to read previously entered text into the editor unless I select the whole text.
The CodeMirror editor has a <textarea> tag, but setting the contenteditable attribute to true doesn't seem to change this behaviour.
Also, I can't figure out how to save the note without visually clicking the save button of the CodeMirror editor, but @zersiax your screen reader can detect the save button, the bold button and the other ones doesn't it?
What screen reader, web browser and operative system are you using?

**Comment by zersiax** [comment]
@ddoyaguez that is the problem exactly. You can type in the text field but not review your text unless you leave the text field and read the, DOM-technically, next items, which contain both the text in the edit field as well as the line and word count. 
I could have sworn I saw the save button before, but perhaps that was on the create note screen where the editor doesn't show up and you can just fill in title and tags.

As for contentEditable not working, I think that if you go the contentEditable or textarea route, you need to create it from zilch and just have a config option to use the screenreader-friendly widget or the regular one. This is why I proposed actually switching out the editor for a more accessible alternative as a first option.
I am using NVDA on Windows at present, have not tested with Orca.

**Comment by Uzay-G** [comment]
Thanks for bringing this up, I'll see what I can do to fix this and maybe switch editors if need be :)

**Comment by ddoyaguez** [comment]
I think I understand. 

easyMDE has a built-in function to revert to the initial textarea, preserving the text that has been introduced.
What about adding a button only detected by screen readers with the absolute positioning technique [as recommended by WebAIM](https://webaim.org/techniques/css/invisiblecontent/#techniques). On pressing that button, the easyMDE editor would return to the default textarea with the contentEditable attribute set to true. Additionally, another save button could be unhidden when the revert button is pressed, so we make sure that the screen reader is able to detect it.

What I don't know is if it would be necessary to add an aria-live attribute to the container div so the screen reader catches this change.

**Comment by zersiax** [comment]
I think that would be excessive. Using role=alert on an invisible div with just a short message is probably better. "screenreader mode" , " normal mode". That , or make the element to enable the screenreader version a checkbox. If the box is checked, editor is in screenreader mode. Adding aria-live to the container might start reading the entire contents of the editor on change , I'm not sure what it does with contents within an input within a container like that :)
If you can set those proposed changes up on a test server I can take a look to see if it's better, don't really have the time currently to check out the code myself and play around with it or I'd make a pull request :)

**Comment by ddoyaguez** [comment]
Okay, thank you for the suggestions!. I'm taking the suggested checkbox approach. I will post here a link to a test server when it is ready. I have also noticed that the delete button, next to the edit button, has no label attached to it, so probably screen reader users can't notice it. @Uzay-G may I add a label there too?

**Comment by Uzay-G** [comment]
Yeah that'd be perfect :)

**Comment by zersiax** [comment]
the button is visible, but unlabeled as you stated. so it just reports as " button"  rather than " delete button"  as it were. aria-label on the button element will fix, or just add text to the button next to the icon.

**Comment by ddoyaguez** [comment]
Hi. I added text to the button next to the icon. Test server with the implemented changes is available from [this link](http://ddoyaguez.ddns.net). Username is admin, and password is admin123.

**Comment by zersiax** [comment]
I just tested the test server.
The button works great, it now reads properly as Delete. The checkbox to enable screenreader mode works as well.
However, the editor appears to still be inaccessible.
Please have a look at the link below, which explains why codeMirror-based editors currently aren't accessible. It seems that altering the way easyMD works won't fix things because the problem is further down the line. This page explains why/how and shows some alternatives. The examples are admittedly old but might shed some light on the issue:
https://bgrins.github.io/codemirror-accessible/

**Comment by ddoyaguez** [comment]
Thank you for testing. With the current method I am destroying the easyMDE editor when the checkbox is activated, and replacing it with a textarea with contentEditable attribute set to true. It indeed solved the issue in Linux with Orca. So it is no longer an easyMDE problem.
I'm installing NVDA on Windows to do further research on this problem.

**Comment by zersiax** [comment]
That is very odd, it felt to me like the easyMD editor was still in place, which makes me wonder if something may have broken on me while I was testing. I tried rechecking the demo server but it is no longer active :) If this does work for you on NVDA , it may have been a glitch on my end. In particular, try narrowing left and right through the entered text, that in particular gave me issues :)

**Comment by ddoyaguez** [comment]
Hi!. I have tested it with NVDA, Windows 10 and Firefox and it works fine for me (if I'm right about what "works" is :) ). I have added an aria-label attribute to the textarea for easier identification of the editor. One thing I have noticed is that when the note is saved (by pressing the save button next to the textarea editor) no confirmation is given. Any suggestions on how to improve that?
Test server is now online, I'm sorry about the downtimes, I have configured it in a VM in my home machine. 
Thank you very much for your patience.

**Comment by zersiax** [comment]
@ddoyaguez I see the issue :) It works fine in Firefox, it breaks in Chrome. You can get rid of the contentEditable, that fixes it in Chrome and given this is already a textarea, we don't need it here.
That does mean that we won't have an easy WYSIWYG way for screenreaders to mark stuff up, but I'd say being able to write and review what you're writing is already a major win.
As for announcing saved state, you could make a little invisible div or span that sends alerts to the screenreader via ARIA, see:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role

**Comment by ddoyaguez** [comment]
Great! Yes, I suppose the optimal solution would be to make easyMDE completely accesible, but until that I'm glad this could help. 
I have implemented the changes following your suggestions (thank you again). Everything seems to work great on my side, the span seems to send alerts correctly. Test server is online if you want to test it. If everything works fine on your side, I will prepare the pull request with the changes.

**Comment by zersiax** [comment]
Just checked, and this looks good :)

**Comment by Uzay-G** [comment]
Ok this should be fixed in #208, @zersiax I'll tell you when we publish a new release with the change :)
